### PR TITLE
chore(e2e): enable e2e tests to run directly against Atlas

### PR DESCRIPTION
**Pull request by syn-zhu**
## Description
https://jira.mongodb.org/browse/CLOUDP-274595

Enable running compass E2E tests directly against a deployed Atlas environment.

Additionally, enable specifying a test filter as env var to support only running a subset of tests.

Example usage:

```
export CONNECTION_STRING_1='mongodb+srv://<username>:<password>@connection-1.dkur9.mongodb-dev.net/?retryWrites=true&w=majority&appName=connection-1'
export CONNECTION_NAME_1=connection-1
export CONNECTION_STRING_2='mongodb+srv://<username>:<password>@connection-2.dkur9.mongodb-dev.net/?retryWrites=true&w=majority&appName=connection-2'
export CONNECTION_NAME_2=connection-2
export ATLAS_DOMAIN=cloud-dev.mongodb.com
export ATLAS_GROUP_ID=66ce1716209aca269796919b
export COOKIES_FILE=/Users/siyanzhu/compass/packages/compass-e2e-tests/cloud-dev.mongodb.com_cookies.json
export E2E_TEST_FILTER=collection-rename

npm run test-web  
```

^This would run the `collection-rename` test against cloud-dev, for group `66ce1716209aca269796919b`.

The current implementation assumes that the actual clusters have already been created beforehand. Later on, we will do this step using the MMS e2e utilities.

Currently, the collection-rename test has been tried and is working. Will do each of the remaining tests incrementally, and make any additional changes along the way.

### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-e2e-tests/helpers/compass.ts [hunk 1]
```diff
@@ -787,7 +792,43 @@ export async function startBrowser(
...
+    // Navigate to a 404 page to set cookies
+    await browser.navigateTo(`https://${ATLAS_DOMAIN}/404`);
+
+    const cookiesFile = process.env.COOKIES_FILE;
+    if (!cookiesFile) {
+      throw new Error(
+        'ATLAS_DOMAIN is set but COOKIES_FILE is not. Please set COOKIES_FILE to the path of the cookies file.'
+      );
+    }
+    const cookies = JSON.parse(await fs.readFile(cookiesFile, 'utf8'));
```
Comment (syn-zhu) [inline packages/compass-e2e-tests/helpers/compass.ts:806]: Eventually, the plan is to package `npm run test-web` as a container image, which can be pulled in MMS e2e task.

MMS already has a bunch of utilities for e2e test setup, in particular:
* Creating a new randomly generated test user
* logging in as that user
* Creating a new group
* Creating new clusters in that group

Thus, we probably don't need to reimplement all of those here. Instead, we can do those setup steps first, and then just save the cookies that are needed for authentication. Then, when we run the compass e2e test container, we just pass in the cookies, the group, and the names of the created clusters.

File: packages/compass-e2e-tests/helpers/compass.ts [hunk 2]
```diff
@@ -787,7 +792,43 @@ export async function startBrowser(
...
+
+    for (const cookie of cookies) {
+      if (
+        cookie.name.includes('mmsa-') ||
+        cookie.name.includes('mdb-sat') ||
+        cookie.name.includes('mdb-srt')
```
Comment (syn-zhu) [inline packages/compass-e2e-tests/helpers/compass.ts:812]: These are the cookies that are relevant for authenticating.

https://github.com/10gen/mms/blob/6d27992a6ab9ab31471c8bcdaa4e347aa39f4013/server/src/features/com/xgen/svc/cukes/helpers/Client.java#L122-L130

File: packages/compass-e2e-tests/helpers/compass.ts (hunk 1 above)
Comment (gribnoysup) [inline packages/compass-e2e-tests/helpers/compass.ts:806]: Can you add some type definitions for this file?

File: packages/compass-e2e-tests/index.ts [hunk 3]
```diff
@@ -118,8 +121,10 @@ async function setup() {
 
   fs.mkdirSync(LOG_PATH, { recursive: true });
 
-  debug('Getting mongodb server info');
-  updateMongoDBServerInfo();
+  if (!ATLAS_DOMAIN) {
+    debug('Getting mongodb server info');
+    updateMongoDBServerInfo();
```
Comment (gribnoysup) [inline packages/compass-e2e-tests/index.ts:126]: Various checks in tests rely on this information existing, you can't just skip it

File: packages/compass-e2e-tests/index.ts [hunk 4]
```diff
@@ -271,15 +279,20 @@ async function main() {
...
   // will also get the slow first run experience for no good reason unless it is
   // the time-to-first-query.ts test.
   // So yeah.. this is a bit of a micro optimisation.
-  const tests = [FIRST_TEST, ...rawTests.filter((t) => t !== FIRST_TEST)];
+  for (let i = 0; i < tests.length; ++i) {
+    if (tests[i] === FIRST_TEST) {
+      [tests[i], tests[0]] = [tests[0], tests[i]];
+      break;
+    }
+  }
```
Comment (gribnoysup) [inline packages/compass-e2e-tests/index.ts:295]: Why is this changed?

File: packages/compass-e2e-tests/tests/collection-rename.test.ts [hunk 5]
```diff
@@ -84,7 +85,10 @@ describe('Collection Rename Modal', () => {
   before(async function () {
     compass = await init(this.test?.fullTitle());
     browser = compass.browser;
-    await browser.setupDefaultConnections();
+    // no need to setup connections if we are running against Atlas
+    if (!ATLAS_DOMAIN) {
+      await browser.setupDefaultConnections();
```
Comment (gribnoysup) [inline packages/compass-e2e-tests/tests/collection-rename.test.ts:90]: I guess it's just temporarily only in this file? You'd need to do this in almost every test file and the pattern is to try to keep the test code as similar as possible (not to mention that changing this logic if we start adding things like that everywhere would be very hard)

File: packages/compass-e2e-tests/helpers/compass.ts [hunk 6]
```diff
@@ -46,6 +46,8 @@ let MONGODB_USE_ENTERPRISE =
 
 // should we test compass-web (true) or compass electron (false)?
 export const TEST_COMPASS_WEB = process.argv.includes('--test-compass-web');
+export const ATLAS_DOMAIN = process.env.ATLAS_DOMAIN;
```
Comment (gribnoysup) [inline packages/compass-e2e-tests/helpers/compass.ts:49]: For all these new variables you're adding, can we make the names more distinct to indicate the use-case better? Especially the CONNECTION_ ones, a lot of things rely on these, so just allowing arbitrary overrides without context is not great

File: packages/compass-e2e-tests/index.ts (hunk 4 above)
Comment (syn-zhu) [reply packages/compass-e2e-tests/index.ts:295]: Because I added the test filtering feature, and this line previously was always including the FIRST_TEST.

File: packages/compass-e2e-tests/index.ts (hunk 3 above)
Comment (syn-zhu) [reply packages/compass-e2e-tests/index.ts:126]: Ah I think I need to update the function to use the variable instead of hardcoded address like it currently does

File: packages/compass-e2e-tests/helpers/compass.ts (hunk 6 above)
Comment (syn-zhu) [reply packages/compass-e2e-tests/helpers/compass.ts:49]: Sure, do you have some concrete suggestions? 

For the connection_ ones in particular, do you just mean adding ATLAS_ to the name or something

File: packages/compass-e2e-tests/helpers/compass.ts (hunk 1 above)
Comment (syn-zhu) [reply packages/compass-e2e-tests/helpers/compass.ts:806]: You mean like, for the cookies object?

File: packages/compass-e2e-tests/tests/collection-rename.test.ts (hunk 5 above)
Comment (syn-zhu) [reply packages/compass-e2e-tests/tests/collection-rename.test.ts:90]: Well, in Atlas when you create new clusters, then the connections start showing up automatically in Data Explore. So there is no extra step needed to "setup" the connections via the connection creation modal.

So, I actually do think we would skip this step for every test *if* we're running against Atlas. Are there any concerns about this?

File: packages/compass-e2e-tests/index.ts (hunk 3 above)
Comment (gribnoysup) [reply packages/compass-e2e-tests/index.ts:126]: Yeah, probably can use DEFAULT_CONNECTION_STRING_1 there insteasd of just port, this is aligned more with what's expected

File: packages/compass-e2e-tests/index.ts (hunk 4 above)
Comment (gribnoysup) [reply packages/compass-e2e-tests/index.ts:295]: Ah, gotcha, was kinda hard to understand from the code. Maybe better to be a bit more explicit about it?

```suggestion
  const shouldRunAllTests = e2eTestFilter === '*';
  const tests = shouldRunAllTests ? <logic to move FIRST_TEST to the front> : <just exactly what glob returned>
```

File: packages/compass-e2e-tests/index.ts (hunk 3 above)
Comment (syn-zhu) [reply packages/compass-e2e-tests/index.ts]: [9585f50](https://github.com/mongodb-js/compass/pull/6381/commits/9585f5043155bec8d7c0ff6d30e5f409d809556c)

File: packages/compass-e2e-tests/index.ts (hunk 4 above)
Comment (syn-zhu) [reply packages/compass-e2e-tests/index.ts:293]: Sure, will update

File: packages/compass-e2e-tests/helpers/compass.ts (hunk 1 above)
Comment (gribnoysup) [reply packages/compass-e2e-tests/helpers/compass.ts:807]: Yes, something like:

```suggestion
    type StoredAtlasCloudCookies = {...some shape}[]
    const cookies: StoredAtlasCloudCookies = JSON.parse(await fs.readFile(cookiesFile, 'utf8'));
```

We probably don't need any strict validation, but without the type it's unclear what's the expected shape for the content of the COOKIES_FILE file and taking into account that it will be generated in a completely separate repo, it's probably our best way of documenting this

File: packages/compass-e2e-tests/helpers/compass.ts (hunk 6 above)
Comment (gribnoysup) [reply packages/compass-e2e-tests/helpers/compass.ts:49]: Yeah, a common prefix we can stick on all of them is probably the way to go, something like `ATLAS_CLOUD_EXTERNAL_URL_TESTING_`, just so it's clear it's specifically for this "testing a deployed atlas cloud" scenario, not a generic way to control these variables

File: packages/compass-e2e-tests/tests/collection-rename.test.ts (hunk 5 above)
Comment (gribnoysup) [reply packages/compass-e2e-tests/tests/collection-rename.test.ts:90]: If switch statements are generic like that, we add them in the helper methods themselves, not in the test case, so again, if this is just temporary because that's the only test case you are adding, that's fine, but for follow-up PRs we'd need to move this logic out of test cases

File: packages/compass-e2e-tests/helpers/compass.ts (hunk 6 above)
Comment (gribnoysup) [reply packages/compass-e2e-tests/helpers/compass.ts:49]: Or rather `TEST_ATLAS_CLOUD_EXTERNAL_URL_` to match existing ones

File: packages/compass-e2e-tests/tests/collection-rename.test.ts (hunk 5 above)
Comment (syn-zhu) [reply packages/compass-e2e-tests/tests/collection-rename.test.ts:90]: Maybe I can add this check into the function itself? Would that be better?

Comment (syn-zhu) [reply packages/compass-e2e-tests/tests/collection-rename.test.ts:90]: Oh oops, just saw you already suggested that lol

File: packages/compass-e2e-tests/index.ts (hunk 4 above)
Comment (syn-zhu) [reply packages/compass-e2e-tests/index.ts:293]: Actually, looking at this again, I think it's more like we want to move time-to-first-query.test.ts to the front only if it is one of the tests we will run (it may still be included if a filter is used).

So technically, we need to check if it's in the full array, and that point we are already iterating the array, so we probably don't need to disambiguate the two cases here.

File: packages/compass-e2e-tests/helpers/compass.ts [hunk 7]
```diff
@@ -106,7 +113,8 @@ export function updateMongoDBServerInfo() {
         'server-info',
         '--',
         '--connectionString',
-        `mongodb://127.0.0.1:${String(MONGODB_TEST_SERVER_PORT)}`,
+        process.env.TEST_ATLAS_CLOUD_EXTERNAL_CONNECTION_STRING_1 ||
+          `mongodb://127.0.0.1:${String(MONGODB_TEST_SERVER_PORT)}`,
```
Comment (gribnoysup) [inline packages/compass-e2e-tests/helpers/compass.ts:117]:
```suggestion
        DEFAULT_CONNECTION_STRING_1
```

File: packages/compass-e2e-tests/helpers/compass.ts [hunk 8]
```diff
@@ -787,7 +805,45 @@ export async function startBrowser(
     ...webdriverOptions,
     ...wdioOptions,
   })) as CompassBrowser;
-  await browser.navigateTo('http://localhost:7777/');
+
+  if (TEST_ATLAS_CLOUD_EXTERNAL_URL) {
+    // Navigate to a 404 page to set cookies
+    await browser.navigateTo(`https://${TEST_ATLAS_CLOUD_EXTERNAL_URL}/404`);
```
Comment (gribnoysup) [inline packages/compass-e2e-tests/helpers/compass.ts:811]: Why are wqe doing this if we're setting cookies from a file below?

File: packages/compass-e2e-tests/helpers/compass.ts [hunk 9]
```diff
@@ -787,7 +805,45 @@ export async function startBrowser(
...
+
+    const cookiesFile = process.env.COOKIES_FILE;
```
Comment (gribnoysup) [inline packages/compass-e2e-tests/helpers/compass.ts:813]:
```suggestion
    const cookiesFile = process.env.TEST_ATLAS_CLOUD_EXTERNAL_COOKIES_FILE;
```

File: packages/compass-e2e-tests/helpers/compass.ts (hunk 2 above)
Comment (gribnoysup) [reply packages/compass-e2e-tests/helpers/compass.ts:827]: Can we add this link as a comment here so that we have the reference and documentation in the code?

File: packages/compass-e2e-tests/index.ts (hunk 4 above)
Comment (gribnoysup) [reply packages/compass-e2e-tests/index.ts:293]: Sure, if the intent is to be able to pick up more than one file, this makes sense. Can we rewrite this using a `sort` then instead of in place swapping so that the code is readable and the intent is clear?

File: packages/compass-e2e-tests/helpers/compass.ts [hunk 10]
```diff
@@ -787,7 +805,45 @@ export async function startBrowser(
...
+        cookie.name.includes('mdb-srt')
+      ) {
+        await browser.setCookies({
+          name: cookie.name,
+          value: cookie.value,
+          domain: cookie.domain,
+          path: cookie.path,
+          secure: cookie.secure,
+          httpOnly: cookie.httpOnly,
+        });
```
Comment (gribnoysup) [inline packages/compass-e2e-tests/helpers/compass.ts:836]: `setCookies` [takes an array](https://v8.webdriver.io/docs/api/browser/setCookies#parameters), you don't need to do this in sequence

File: packages/compass-e2e-tests/helpers/compass.ts (hunk 7 above)
Comment (syn-zhu) [reply packages/compass-e2e-tests/helpers/compass.ts:117]: DEFAULT_CONNECTION_STRING_1 has the /test suffix in its definition, but the one here doesn't. 

I think that's supposed to refer to the auth DB that is used? if you think it won't make a difference then I can change it to what you suggested.

File: packages/compass-e2e-tests/helpers/compass.ts (hunk 8 above)
Comment (gribnoysup) [reply packages/compass-e2e-tests/helpers/compass.ts:811]: Ah, okay, I thought we're going on this page so that the set-cookie header from the backend can set some cookies, not so that the `browser.setCookies` call can work, got it!

**Review by gribnoysup — APPROVED** [review]
We should rename the rest of some specific Atlas Cloud test related variables (cookie file is the only one left I think, at least the one I noticed on re-review), otherwise a few nits, but looks good to merge
